cabby/model/landmark_recognition/dataset_bert.py

```diff
@@ -0,0 +1,180 @@
+    basic_tokenization = [
+      basic_tokenize_and_align_labels(sent, labs)
+      for sent, labs in zip(data.instructions.tolist(), data.entity_span)
+    ]
+    self.inputs = [bert_tokenize_and_align_labels(sent, labs) for sent, labs in basic_tokenization]
+    self.sent = data.instructions.tolist()
+
+
+  def __getitem__(self, idx: int):
+    '''Supports indexing such that TextGeoDataset[i] can be used to get
+    i-th sample.
+    Arguments:
+      idx: The index for which a sample from the dataset will be returned.
+    Returns:
+      A single sample including text, the tokenization of the text and
+      the corresponding labels.
+    '''
+
+    input = {k: torch.tensor(v) for k, v in self.inputs[idx].items()}
+    input['instructions'] = self.sent[idx]
+    return input
+
+  def __len__(self):
+    return len(self.inputs)
+
+def create_dataset(
+  data_dir: str,
+  region: str,
+  s2level: int,
+) -> Tuple[EntityRecognitionSplit, EntityRecognitionSplit, EntityRecognitionSplit]:
+  '''Loads data and creates datasets and train, validate and test sets.
+  Arguments:
+    data_dir: The directory of the data.
+    region: The region of the data.
+    s2level: The s2level of the cells.
+  Returns:
+    The train, validate and test sets.
+  '''
+  rvs_dataset = datasets.RVSDataset(data_dir, s2level, region)
+
+  train_dataset = EntityRecognitionSplit(
+    rvs_dataset.train)
+  logging.info(
+    f"Finished to create the train-set with {len(train_dataset)} samples")
+  val_dataset = EntityRecognitionSplit(
+    rvs_dataset.valid)
+  logging.info(
+    f"Finished to create the valid-set with {len(val_dataset)} samples")
+  test_dataset = EntityRecognitionSplit(
+    rvs_dataset.test)
+  logging.info(
+    f"Finished to create the test-set with {len(test_dataset)} samples")
+
+  return train_dataset, val_dataset, test_dataset
+
+def basic_tokenize_and_align_labels(
+  sentence: str, text_labels: Dict[str,Tuple[int, int]]
+) :
+  '''Tokenize sentence and preserve the labels,
+  such that for each token there is a label.
+  Arguments:
+    sentence: the sentence to tokenize.
+    text_labels: the dictionary containing the landmarks (keys) and span (values).
+  Returns:
+    A List of the tokens and their corresponding labels.
+  '''
+
+  # Basic tokenization of the sentence.
+  sentence_words = word_tokenize(sentence)
+
+  # Create labels for each token according to the dictionary of entities and spans.
+  labels = []
+  spans = sorted(list(text_labels.values()))
+
+  cur_index = 0
+  span = spans.pop(0)
+  start, end = span[0], span[1]
+  for word in sentence_words:
+    if cur_index >= start:
+      labels.append(1)
+    else:
+      labels.append(0)
+
+    cur_index += len(word)
+    if cur_index<len(sentence) and sentence[cur_index] == ' ':
+      cur_index+=1
+    if cur_index >= end:
+      if len(spans)>0:
+        span = spans.pop(0)
+        start, end = span[0], span[1]
+      else:
+        start = len(sentence)+1

```

Comment:
This looks like a similar loop from before (maybe exact duplicate?). Can you abstract this in a function and write tests for it?

Reply:
I will remove dataset.py